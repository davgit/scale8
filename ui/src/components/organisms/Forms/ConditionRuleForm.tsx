import { FC, useEffect, useState } from 'react';
import { Box, Checkbox, FormControlLabel } from '@mui/material';
import { ConditionType, InputType, VarType } from '../../../gql/generated/globalTypes';
import {
    ConditionRuleFormProps,
    ConditionRuleValues,
} from '../../../hooks/form/useConditionRuleForm';
import { FormElementWithInputType } from '../../atoms/InputTypes/FormElementWithInputType';
import DrawerFormLayout from '../../molecules/DrawerFormLayout';
import { matchConditionValues } from '../../../utils/MatchConditionValues';
import { extractDataMapDefaultValue } from '../../../utils/MappedPlatformElementUtils';
import { buildConditionName } from '../../../dialogPages/tagManager/app/trigger/ConditionRuleUpdate';
import { snakeToTitleCase } from '../../../utils/TextUtils';
import { platformDataMapsToSelectValues } from '../../../utils/DataContainersUtils';
import { DialogFormContextProvider } from '../../../context/DialogFormContext';
import { DialogFormFilteredSelects } from '../../atoms/DialogFormInputs/DialogFormFilteredSelects';
import { DialogFormTextInput } from '../../atoms/DialogFormInputs/DialogFormTextInput';
import { DialogPlainAlert } from '../../atoms/DialogFormInputs/DialogPlainAlert';
import { DialogFormSelect } from '../../atoms/DialogFormInputs/DialogFormSelect';
import { DialogFormTextAreaInput } from '../../atoms/DialogFormInputs/DialogFormTextAreaInput';

const ConditionRuleForm: FC<ConditionRuleFormProps> = (props: ConditionRuleFormProps) => {
    const [useCustom, setUseCustom] = useState(false);
    const [generateName, setGenerateName] = useState(
        props.generateName === undefined ? true : props.generateName,
    );

    const { values, currentDataContainer, handleChange, dataContainers } = props;

    useEffect(() => {
        if (
            currentDataContainer &&
            currentDataContainer.allow_custom &&
            currentDataContainer.platform_data_maps.length < 1
        ) {
            setUseCustom(true);
        } else {
            setUseCustom(false);
        }
    }, [currentDataContainer]);

    const noContainers = props.dataContainers.length < 1;

    const notAvailable =
        currentDataContainer !== undefined &&
        !currentDataContainer.allow_custom &&
        currentDataContainer.platform_data_maps.length === 0;

    const hasDataMaps =
        currentDataContainer !== undefined && currentDataContainer.platform_data_maps.length > 0;

    const hasDataMapsAndCustom =
        currentDataContainer !== undefined &&
        currentDataContainer.platform_data_maps.length > 0 &&
        currentDataContainer.allow_custom;

    const ready = !notAvailable && !noContainers && props.currentDataContainer;

    useEffect(() => {
        const currentName = buildConditionName(values, dataContainers, currentDataContainer);

        if (ready && generateName && values.name !== currentName) {
            handleChange('name', currentName);
        }
    }, [ready, values, currentDataContainer, handleChange, generateName, dataContainers]);

    const isRawObject =
        props.currentDataElement !== undefined &&
        props.currentDataElement.var_type === VarType.OBJECT &&
        (props.currentDataElement.child_platform_data_maps ?? []).length === 0;

    const findConditionInputType = (): InputType | 'Custom' => {
        if (props.currentDataElement === undefined || isRawObject) {
            return 'Custom';
        }

        return props.currentDataElement.input_type;
    };

    const conditionInputType = findConditionInputType();

    const filteredConditions = matchConditionValues.filter((matchConditionValue) =>
        matchConditionValue.types.includes(conditionInputType),
    );

    const noConditions = filteredConditions.length === 0;

    return (
        <DialogFormContextProvider<ConditionRuleValues> formProps={props}>
            <DrawerFormLayout
                {...props}
                submitDisable={props.isSubmitting || !ready || noConditions}
            >
                {noContainers ? (
                    <small>There are no platforms with data containers available.</small>
                ) : (
                    <DialogFormFilteredSelects
                        disabled={!!props.update}
                        label="Data Container"
                        initialFilterValue={props.initialPlatformId}
                        name="dataContainerId"
                        values={dataContainers}
                        filterLabel="Platform"
                        missingSubMessage="There are no data containers available in this platform."
                    />
                )}
                {currentDataContainer && (
                    <Box
                        component="small"
                        sx={{
                            width: '100%',
                            margin: (theme) => theme.spacing(0, 0, 2),
                        }}
                    >
                        {currentDataContainer.description}
                    </Box>
                )}

                {notAvailable && (
                    <small>It is not possible to use this data container in a condition.</small>
                )}

                {hasDataMapsAndCustom && (
                    <FormControlLabel
                        control={
                            <Checkbox
                                name="useCustom"
                                checked={useCustom}
                                onChange={(event) => {
                                    if (event.target.checked) {
                                        // force values reset by pretend to change data container
                                        handleChange(
                                            'dataContainerId',
                                            props.values.dataContainerId,
                                        );
                                        setUseCustom(true);
                                    } else {
                                        // force values reset by pretend to change data container
                                        handleChange(
                                            'dataContainerId',
                                            props.values.dataContainerId,
                                        );
                                        setUseCustom(false);
                                    }
                                }}
                                color="primary"
                            />
                        }
                        label="Use Custom Data Element"
                    />
                )}

                {useCustom && <DialogFormTextInput name="match" label="Custom Data Element" />}

                {currentDataContainer && hasDataMaps && !useCustom && (
                    <DialogFormSelect
                        label="Data Element"
                        name="matchId"
                        values={platformDataMapsToSelectValues(
                            currentDataContainer.platform_data_maps,
                        )}
                    />
                )}
                {noConditions ? (
                    <DialogPlainAlert>
                        It is not possible to use this data element in a condition.
                    </DialogPlainAlert>
                ) : (
                    (props.currentDataElement !== undefined || useCustom) && (
                        <>
                            {props.currentDataElement && (
                                <DialogPlainAlert>
                                    ({snakeToTitleCase(props.currentDataElement.var_type)}){' '}
                                    {props.currentDataElement.description}
                                </DialogPlainAlert>
                            )}
                            {isRawObject && <DialogFormTextInput name="matchKey" label="Key" />}
                            <DialogFormSelect
                                label="Condition"
                                name="matchCondition"
                                values={filteredConditions.map((matchConditionValue) => ({
                                    key: matchConditionValue.condition,
                                    text: `${matchConditionValue.symbol} (${matchConditionValue.text})`,
                                }))}
                            />
                            {props.values.matchCondition !== ConditionType.IS_DEFINED &&
                                props.values.matchCondition !== ConditionType.IS_NOT_DEFINED && (
                                    <FormElementWithInputType
                                        {...{
                                            label: 'Value',
                                            name: 'dataMapValue',
                                            values: [props.values.dataMapValue],
                                            inputType: isRawObject
                                                ? InputType.TEXT
                                                : props.currentDataElement?.input_type ??
                                                  InputType.TEXT,
                                            optionValues:
                                                props.currentDataElement?.option_values ?? [],
                                            setValue: (v) => {
                                                props.handleChange('dataMapValue', v);
                                            },
                                            removeArrayElement: () => {
                                                // Flatten
                                            },
                                            addArrayElement: () => {
                                                // Flatten
                                            },
                                            consentPurposes: props.consentPurposes,
                                            consentVendors: props.consentVendors,
                                            defaultValue:
                                                props.currentDataElement !== undefined &&
                                                props.currentDataElement !== null
                                                    ? extractDataMapDefaultValue(
                                                          props.currentDataElement,
                                                      )
                                                    : null,
                                            onBlur: props.handleBlur,
                                            errors: props.errors.dataMapValue
                                                ? new Map([[0, props.errors.dataMapValue]])
                                                : new Map(),
                                            disabled: false,
                                            required: true,
                                            flatten: true,
                                            className: 'DialogFormField',
                                        }}
                                    />
                                )}
                        </>
                    )
                )}
                {ready && !noConditions && (
                    <>
                        <FormControlLabel
                            sx={{ marginBottom: (theme) => theme.spacing(3) }}
                            control={
                                <Checkbox
                                    name="generateName"
                                    checked={generateName}
                                    onChange={(event) => {
                                        setGenerateName(event.target.checked);
                                    }}
                                    color="primary"
                                />
                            }
                            label="Generate Name"
                        />
                        {!generateName && <DialogFormTextInput name="name" label="Name" />}
                    </>
                )}
                <DialogFormTextAreaInput name="comments" label="Comments" optional />
            </DrawerFormLayout>
        </DialogFormContextProvider>
    );
};

export default ConditionRuleForm;
