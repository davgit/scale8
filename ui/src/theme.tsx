import { createTheme, adaptV4Theme } from '@mui/material/styles';
import { getProductSection, ProductSectionKey } from './containers/SectionsDetails';
import { grey } from '@mui/material/colors';

declare module '@mui/material/styles/createPalette' {
    interface Palette {
        adminColor: Palette['primary'];
        dataManagerColor: Palette['primary'];
        tagManagerColor: Palette['primary'];
        commonColor: Palette['primary'];
    }
    interface PaletteOptions {
        adminColor: PaletteOptions['primary'];
        dataManagerColor: PaletteOptions['primary'];
        tagManagerColor: PaletteOptions['primary'];
        commonColor: PaletteOptions['primary'];
    }
}

// A custom theme for this app
const theme = createTheme(
    adaptV4Theme({
        palette: {
            primary: {
                main: '#25302e',
            },
            secondary: {
                main: grey['700'],
            },
            error: {
                main: '#c63d51',
            },
            info: {
                main: '#444444',
            },
            success: {
                main: '#c6e54b',
            },
            background: {
                default: '#ffffff',
            },
            adminColor: {
                main: getProductSection(ProductSectionKey.admin).color,
            },
            dataManagerColor: {
                main: getProductSection(ProductSectionKey.dataManager).color,
            },
            tagManagerColor: {
                main: getProductSection(ProductSectionKey.tagManager).color,
            },
            commonColor: {
                main: getProductSection(ProductSectionKey.global).color,
            },
        },
        typography: {
            fontFamily: ['Source Sans Pro', 'Helvetica Neue', 'Arial', 'sans-serif'].join(','),
            fontSize: 14,
            body1: {
                fontSize: 14,
            },
        },
        overrides: {
            MuiTableCell: {
                head: {
                    fontWeight: 600,
                },
            },
            MuiSvgIcon: {
                root: {
                    fontSize: '1.25rem',
                },
            },
            MuiButton: {
                root: {
                    fontWeight: 600,
                },
            },
            MuiTooltip: {
                tooltip: {
                    fontSize: '.8rem',
                },
            },
            MuiDialogActions: {
                spacing: {
                    '& > :not(:first-child)': {
                        marginLeft: '16px',
                    },
                },
            },
        },
    }),
);

export default theme;
