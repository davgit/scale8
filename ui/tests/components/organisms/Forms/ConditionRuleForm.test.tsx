import { render } from '@testing-library/react';
import { ThemeProvider } from '@mui/material';
import theme from '../../../../src/theme';
import { FC } from 'react';
import ConditionRuleForm from '../../../../src/components/organisms/Forms/ConditionRuleForm';
import { useMockConditionFormProps } from '../../../componentsMocks/useMockConditionFormProps';

const MockForm: FC = () => {
    const formProps = useMockConditionFormProps();

    return <ConditionRuleForm {...formProps} />;
};

describe('ConditionRuleForm', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });
    it('form render unchanged', () => {
        const { asFragment } = render(
            <ThemeProvider theme={theme}>
                <MockForm />
            </ThemeProvider>,
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
