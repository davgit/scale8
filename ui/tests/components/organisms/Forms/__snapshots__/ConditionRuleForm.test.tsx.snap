// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConditionRuleForm form render unchanged 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-s92abg"
  >
    <div
      class="MuiBox-root css-70qvj9"
    >
      <div
        class="MuiBox-root css-1isemmb"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-erxxaa-MuiSvgIcon-root"
            data-testid="CloseIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root css-uix0gv"
      >
        Submit Condition
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-yn9cjk"
    >
      <div
        class="MuiBox-root css-1676va5"
      >
        <form
          class="MuiBox-root css-mjj15b"
        >
          <div
            class="MuiFormControl-root DialogFormField css-1nrlq1o-MuiFormControl-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error Mui-required css-a6y1oi-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
            >
              Data Container
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk Mui-error css-b4y4zd-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl MuiSelect-root css-1ejkwvn-MuiInputBase-root-MuiInput-root"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="mui-component-select-dataContainerId"
                class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
                id="mui-component-select-dataContainerId"
                role="button"
                tabindex="0"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                name="dataContainerId"
                required=""
                tabindex="-1"
                value=""
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-yqkyv6-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
            <p
              class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium Mui-required css-1sbojr7-MuiFormHelperText-root"
            >
              Please select a value
            </p>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root DialogFormField css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-ip4mx2-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
            >
              Comments
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-1yxqjki-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <textarea
                aria-invalid="false"
                autocomplete="off"
                class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                name="comments"
                style="height: 0px; overflow: hidden;"
              />
              <textarea
                aria-hidden="true"
                class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                tabindex="-1"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-dhh0px"
                >
                  <span>
                    Comments
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiBox-root css-wpygli"
          >
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled css-xcbv7o-MuiButtonBase-root-MuiButton-root"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              Submit Condition
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
